cms/service/EvaluationService.py

```diff
@@ -142,6 +151,19 @@ def execute(self, entries):
                     self._currently_executing = []
                     break
 
+    def enqueue(self, item, priority=None, timestamp=None):
+        if super().enqueue(item, priority, timestamp):
+            # Add the item to the cumulative status dictionary.
+            key = item.short_key()
+            if key in self.queue_status_cumulative:
+                self.queue_status_cumulative[key]["item"]["multiplicity"] += 1
+            else:
+                item_entry = item.to_dict()
+                del item_entry["testcase_codename"]
+                item_entry["multiplicity"] = 1
+                entry = {"item": item_entry, "priority": priority, "timestamp": make_timestamp(timestamp)}
```

Comment:
Actually, I think we always pass all three arguments to `EvaluationExecutor.enqueue`, so `priority` and `timestamp` are never `None`. How about we make the arguments mandatory in the definition of the method?

After reading your comment, I realized that we are implicitly assuming that the priority and timestamp are uniquely determined by the (type, object_id, dataset_id) tuple. Otherwise, it's unclear what we should display in the corresponding line on the overview page.

As far as I see, the timestamp is the time the submission was made, and therefore determined by object_id.

However, the priority can vary: For example, when the evaluation of a testcase fails, it is re-attempted with a lower priority.

How about we add the priority value to the tuple (and perhaps timestamp for safety?), so use (type, object_id, dataset_id, priority, timestamp)? Basically, I think everything that is displayed to the user (except multiplicity) should be part of the key tuple.

Reply:
> How about we make the arguments mandatory in the definition of the method?

This is a nice suggestion. Normally it would be awkward to redefine the prototype to differ from the overridden method, but here usage is contained to one file (sometimes going through `TriggeredService`) and, more so, `EvaluationService.enqueue()` already does the same thing. So we can do it here too.

> How about we add the priority value to the tuple (and perhaps timestamp for safety?), so use (type, object_id, dataset_id, priority, timestamp)?

Indeed, priority could be possibly added. The upside is that it won't merge different items as master does, the downside is that it'll be possible to have multiple entries with the same description text (sans multiplicity) in AWS. You'll need to make `Executor.dequeue()` return a value, though.

About timestamp, if it is already tied to `object_id`, then I don't think it needs to be in the key. You could add an assert on mismatch if you think it is useful.

Reply:
Okay, I made the priority and timestamp arguments mandatory.

I also added priority to the tuple. It's true there might then be multiple lines with the same description text. However, the priority is also shown, so I think it should be fairly clear to the admin what's going on. The change should make ordering of the queue entries more consistent. Also, before this change, it was possible that the displayed priority was actually no longer present in the queue at all: Add the same submission with priority 1, then 2. Then remove the first (priority 1) entry. The dictionary will still assume the priority is 1!

As you suggested, I didn't add timestamp to the tuple. I didn't put in an assertion, either, because checking whether floating-point numbers are equal scares me. :)

(By the way, the order lines appear on AWS probably doesn't always agree with the order the items appear in the queue: There might be several lines with the same priority and timestamp, which are shown in arbitrary order. I'm not even certain that items with the same (type,object_id,dataset_id,priority) tuple are guaranteed to be consecutive in the queue...)

Reply:
Thanks, looks good.

> Also, before this change, it was possible that the displayed priority was actually no longer present in the queue at all: Add the same submission with priority 1, then 2. Then remove the first (priority 1) entry. The dictionary will still assume the priority is 1!

Indeed, this is true.